File: app/controllers/iop/transfer/impl/ExternalAttachmentLoaderImpl.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
// SPDX-FileCopyrightText: 2024. The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl;

import com.fasterxml.jackson.databind.JsonNode;
import controllers.iop.transfer.api.ExternalAttachmentLoader;
import java.io.File;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import javax.inject.Inject;
import models.Attachment;
import models.Exam;
import org.apache.pekko.actor.ActorSystem;
import org.apache.pekko.stream.IOResult;
import org.apache.pekko.stream.Materializer;
import org.apache.pekko.stream.javadsl.FileIO;
import org.apache.pekko.stream.javadsl.Source;
import org.apache.pekko.util.ByteString;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.util.ObjectUtils;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.mvc.Http;
import util.config.ConfigReader;
import util.file.FileHandler;

public class ExternalAttachmentLoaderImpl implements ExternalAttachmentLoader {

    @Inject
    private ActorSystem actor;

    @Inject
    private WSClient wsClient;

    @Inject
    private FileHandler fileHandler;

    @Inject
    private ConfigReader configReader;

    private final Logger logger = LoggerFactory.getLogger(ExternalAttachmentLoaderImpl.class);

    @Override
    public CompletableFuture<Void> fetchExternalAttachmentsAsLocal(Exam exam) {
        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
        if (exam.getAttachment() != null) {
            futures.add(createFromExternalAttachment(exam.getAttachment(), "exam", exam.getId().toString()));
        }
        exam
            .getExamSections()
            .stream()
            .flatMap(examSection -> examSection.getSectionQuestions().stream())
            .map(sectionQuestion -> {
                if (
                    sectionQuestion.getEssayAnswer() != null && sectionQuestion.getEssayAnswer().getAttachment() != null
                ) {
                    futures.add(
                        createFromExternalAttachment(
                            sectionQuestion.getEssayAnswer().getAttachment(),
                            "question",
                            sectionQuestion.getId().toString(),
                            "answer",
                            sectionQuestion.getEssayAnswer().getId().toString()
                        )
                    );
                }
                return sectionQuestion.getQuestion();
            })
            .filter(question -> question.getAttachment() != null)
            .distinct()
            .forEach(question ->
                futures.add(
                    createFromExternalAttachment(question.getAttachment(), "question", question.getId().toString())
                )
            );
        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
    }

    @Override
    public CompletableFuture<Void> createExternalAttachment(Attachment attachment) {
        if (attachment == null || ObjectUtils.isEmpty(attachment.getFilePath())) {
            return CompletableFuture.completedFuture(null);
        }
        final URL attachmentUrl;
        try {
            attachmentUrl = parseUrl("/api/attachments/");
        } catch (MalformedURLException e) {
            return CompletableFuture.supplyAsync(() -> {
                throw new RuntimeException(e);
            });
        }
        final WSRequest request = wsClient.url(attachmentUrl.toString());
        return request
            .post("")
            .thenAcceptAsync(response -> {
                final JsonNode json = response.asJson();
                final String externalId = json.get("id").asText();
                attachment.setExternalId(externalId);
                File file = new File(attachment.getFilePath());
                if (!file.exists()) {
                    logger.warn(
                        "Could not find file {} for attachment id {}.",
                        file.getAbsoluteFile(),
                        attachment.getId()
                    );
                    return;
                }
                final WSRequest updateRequest;
                try {
                    updateRequest =
                        wsClient.url(parseUrl("/api/attachments/%s", attachment.getExternalId()).toString());
                } catch (MalformedURLException e) {
                    logger.error("Invalid URL!", e);
                    return;
                }
                final Source<ByteString, CompletionStage<IOResult>> source = FileIO.fromPath(file.toPath());
                final Http.MultipartFormData.FilePart<Source<ByteString, CompletionStage<IOResult>>> filePart =
                    new Http.MultipartFormData.FilePart<>(
                        "file",
                        attachment.getFileName(),
                        attachment.getMimeType(),
                        source
                    );
                Http.MultipartFormData.DataPart dp = new Http.MultipartFormData.DataPart("key", "value");

                updateRequest
                    .put(Source.from(Arrays.asList(filePart, dp)))
                    .thenAccept(wsResponse -> {
                        if (wsResponse.getStatus() != Http.Status.OK) {
                            logger.warn("File upload {} failed!", file.getAbsoluteFile());
                            return;
                        }
                        logger.info("Uploaded file {} for external exam.", file.getAbsoluteFile());
                    });
            })
            .toCompletableFuture();
    }

    @Override
    public CompletableFuture<Void> uploadAssessmentAttachments(Exam exam) {
        List<CompletableFuture<Void>> futures = new ArrayList<>();
        // Create external attachments.
        if (exam.getAttachment() != null) {
            futures.add(createExternalAttachment(exam.getAttachment()));
        }
        exam
            .getExamSections()
            .stream()
            .flatMap(s -> s.getSectionQuestions().stream())
            .flatMap(sq -> {
                List<Attachment> attachments = new ArrayList<>();
                if (sq.getEssayAnswer() != null && sq.getEssayAnswer().getAttachment() != null) {
                    attachments.add(sq.getEssayAnswer().getAttachment());
                }
                if (sq.getQuestion().getAttachment() != null) {
                    attachments.add(sq.getQuestion().getAttachment());
                }
                return attachments.stream();
            })
            .forEach(a -> futures.add(createExternalAttachment(a)));
        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
    }

    private CompletableFuture<Void> createFromExternalAttachment(Attachment attachment, String... pathParams) {
        return CompletableFuture.runAsync(() -> {
            if (ObjectUtils.isEmpty(attachment.getExternalId())) {
                logger.error("Could not find external ID for an attachment");
                return;
            }
            final URL attachmentUrl;
            try {
                attachmentUrl = parseUrl("/api/attachments/%s/download", attachment.getExternalId());
            } catch (MalformedURLException e) {
                throw new RuntimeException("Invalid URL!", e);
            }
            final WSRequest request = wsClient.url(attachmentUrl.toString());
            request
                .stream()
                .thenAccept(response -> {
                    final String filePath = fileHandler.createFilePath(pathParams);
                    response
                        .getBodyAsSource()
                        .runWith(FileIO.toPath(Paths.get(filePath)), Materializer.createMaterializer(actor))
                        .thenAccept(ioResult -> {
                            attachment.setFilePath(filePath);
                            attachment.save();
                            logger.info(
                                "Saved attachment {} locally as # {}",
                                attachment.getExternalId(),
                                attachment.getId()
                            );
                        });
                });
        });
    }

    private URL parseUrl(String format, Object... args) throws MalformedURLException {
        final String path = args.length < 1 ? format : String.format(format, args);
        return URI.create(configReader.getIopHost() + path).toURL();
    }
}

File: app/controllers/iop/transfer/impl/ExternalAttachmentLoaderImpl.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl

import controllers.iop.transfer.api.ExternalAttachmentLoader
import miscellaneous.config.ConfigReader
import miscellaneous.file.FileHandler
import models.attachment.Attachment
import models.exam.Exam
import org.apache.pekko.actor.ActorSystem
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.{FileIO, Source}
import org.apache.pekko.util.ByteString
import play.api.Logging
import play.api.libs.ws.{WSBodyWritables, WSClient}
import play.api.mvc.MultipartFormData

import java.io.File
import java.net.{URI, URL}
import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters.*
import scala.util.Try

class ExternalAttachmentLoaderImpl @Inject() (
    actorSystem: ActorSystem,
    wsClient: WSClient,
    fileHandler: FileHandler,
    configReader: ConfigReader
)(implicit ec: ExecutionContext, mat: Materializer)
    extends ExternalAttachmentLoader
    with WSBodyWritables
    with Logging:

  override def fetchExternalAttachmentsAsLocal(exam: Exam): Future[Unit] =
    val futures = scala.collection.mutable.ListBuffer[Future[Unit]]()

    // Add exam attachment if present
    Option(exam.getAttachment).foreach { attachment =>
      futures += createFromExternalAttachment(attachment, "exam", exam.getId.toString)
    }

    // Process section questions
    exam.getExamSections.asScala
      .flatMap(_.getSectionQuestions.asScala)
      .foreach { sectionQuestion =>
        // Add essay answer attachment if present
        Option(sectionQuestion.getEssayAnswer)
          .flatMap(ea => Option(ea.getAttachment))
          .foreach { attachment =>
            futures += createFromExternalAttachment(
              attachment,
              "question",
              sectionQuestion.getId.toString,
              "answer",
              sectionQuestion.getEssayAnswer.getId.toString
            )
          }
      }

    // Process distinct questions with attachments
    val distinctQuestions = exam.getExamSections.asScala
      .flatMap(_.getSectionQuestions.asScala)
      .map(_.getQuestion)
      .filter(q => q.getAttachment != null)
      .toSeq
      .distinct

    distinctQuestions.foreach { question =>
      futures += createFromExternalAttachment(question.getAttachment, "question", question.getId.toString)
    }

    // Wait for all futures to complete
    Future.sequence(futures.toSeq).map(_ => ())

  override def createExternalAttachment(attachment: Attachment): Future[Unit] =
    Option(attachment).flatMap(a => Option(a.getFilePath).filter(_.nonEmpty)) match
      case None => Future.successful(())
      case Some(_) =>
      parseUrl("/api/attachments/") match
        case None =>
          Future.failed(new RuntimeException("Invalid URL"))
        case Some(attachmentUrl) =>
          val request = wsClient.url(attachmentUrl.toString)
          request.post("").flatMap { response =>
            val json = response.json
            val externalId = (json \ "id").as[String]
            attachment.setExternalId(externalId)

            val file = new File(attachment.getFilePath)
            if !file.exists() then
              logger.warn(
                s"Could not find file ${file.getAbsoluteFile} for attachment id ${attachment.getId}."
              )
              Future.successful(())
            else
              parseUrl("/api/attachments/%s", externalId) match
                case None =>
                  logger.error("Invalid URL!")
                  Future.successful(())
                case Some(updateUrl) =>
                  val updateRequest = wsClient.url(updateUrl.toString)
                  val source = FileIO.fromPath(file.toPath)
                  val filePart = MultipartFormData.FilePart[Source[ByteString, ?]](
                    "file",
                    attachment.getFileName,
                    Option(attachment.getMimeType),
                    source
                  )
                  // Create Source with FilePart (DataPart not needed for Scala WSClient)
                  val partsSource = Source.single(filePart)

                  updateRequest.put(partsSource).map { wsResponse =>
                    if wsResponse.status != play.api.http.Status.OK then
                      logger.warn(s"File upload ${file.getAbsoluteFile} failed!")
                    else
                      logger.info(s"Uploaded file ${file.getAbsoluteFile} for external exam.")
                  }
          }

  override def uploadAssessmentAttachments(exam: Exam): Future[Unit] =
    val futures = scala.collection.mutable.ListBuffer[Future[Unit]]()

    // Create external attachments
    Option(exam.getAttachment).foreach { attachment =>
      futures += createExternalAttachment(attachment)
    }

    // Process section questions
    val attachmentFutures = exam.getExamSections.asScala
      .flatMap(_.getSectionQuestions.asScala)
      .flatMap { sq =>
        val attachments = scala.collection.mutable.ListBuffer[Attachment]()

        Option(sq.getEssayAnswer)
          .flatMap(ea => Option(ea.getAttachment))
          .foreach(attachments += _)

        Option(sq.getQuestion.getAttachment).foreach(attachments += _)

        attachments
      }
      .map(createExternalAttachment)

    futures ++= attachmentFutures

    // Wait for all futures to complete
    Future.sequence(futures.toSeq).map(_ => ())

  private def createFromExternalAttachment(attachment: Attachment, pathParams: String*): Future[Unit] =
    Option(attachment.getExternalId).filter(_.nonEmpty) match
      case None =>
        logger.error("Could not find external ID for an attachment")
        Future.failed(new RuntimeException("Could not find external ID for an attachment"))
      case Some(externalId) =>
        parseUrl("/api/attachments/%s/download", externalId) match
        case None =>
          Future.failed(new RuntimeException("Invalid URL!"))
        case Some(attachmentUrl) =>
          val request = wsClient.url(attachmentUrl.toString)
          request.stream().flatMap { response =>
            val filePath = fileHandler.createFilePath(pathParams*)
            val sink = FileIO.toPath(java.nio.file.Paths.get(filePath))

            response.bodyAsSource.runWith(sink).map { _ =>
              attachment.setFilePath(filePath)
              attachment.save()
              logger.info(
                s"Saved attachment ${attachment.getExternalId} locally as # ${attachment.getId}"
              )
            }
          }

  private def parseUrl(format: String, args: String*): Option[URL] =
    val path = if args.isEmpty then format else format.format(args*)
    Try(URI.create(configReader.getIopHost + path).toURL).toOption

